mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java

```diff
@@ -171,117 +190,207 @@ public void commitJob(JobContext originalContext) throws IOException {
-        .onFailure((file, exc) -> LOG.debug("Failed on to remove directory {} on cleanup job", file, exc))
-        .run(file -> {
-          Path toDelete = new Path(file);
-          FileSystem fs = Util.getFs(toDelete, jobContext.getJobConf());
-          fs.delete(toDelete, true);
-        }, IOException.class);
+    LOG.info("Cleaning for job {} started", jobContext.getJobID());
+    Map<String, String> outputs = SerializationUtil.deserializeFromBase64(jobConf.get(InputFormatConfig.OUTPUT_TABLES));
+
+    for (String table : outputs.keySet()) {
+      String tableLocation = generateJobLocation(outputs.get(table), jobConf, jobContext.getJobID());
+      LOG.info("Cleaning table {} on location: {}", table, tableLocation);
+
+      // Remove the job's temp directory recursively.
+      // Intentionally used foreach on a single item. Using the Tasks API here only for the retry capability.
+      Tasks.foreach(tableLocation)
+          .retry(3)
+          .suppressFailureWhenFinished()
+          .onFailure((file, exc) -> LOG.debug("Failed on to remove directory {} on cleanup job", file, exc))
+          .run(file -> {
+            Path toDelete = new Path(file);
+            FileSystem fs = Util.getFs(toDelete, jobContext.getJobConf());
+            fs.delete(toDelete, true);
+          }, IOException.class);
+    }
+
+    LOG.info("Cleaning for job {} finished", jobContext.getJobID());
+  }
+
+  /**
+   * Executor service for parallel handling of file reads. Should be shared when committing multiple tables.
+   * @param conf The configuration containing the pool size
+   * @return The generated executor service
+   */
+  private static ExecutorService fileReaderExecutor(Configuration conf) {
+    int size = conf.getInt(InputFormatConfig.COMMIT_FILE_THREAD_POOL_SIZE,
+        InputFormatConfig.COMMIT_FILE_THREAD_POOL_SIZE_DEFAULT);
+    return Executors.newFixedThreadPool(
+        size,
+        new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setPriority(Thread.NORM_PRIORITY)
+            .setNameFormat("iceberg-commit-file-pool-%d")
+            .build());
+  }
+
+  /**
+   * Executor service for parallel handling of table manipulation. Could return null, if no parallelism is possible.
+   * @param conf The configuration containing the pool size
+   * @return The generated executor service, or null if executor is not needed.
+   */
+  private static ExecutorService tableExecutor(Configuration conf, int outputSize) {
+    int size = conf.getInt(InputFormatConfig.COMMIT_TABLE_THREAD_POOL_SIZE,
+        InputFormatConfig.COMMIT_TABLE_THREAD_POOL_SIZE_DEFAULT);
+    size = Math.max(outputSize, size);
+    if (size > 1) {
+      return Executors.newFixedThreadPool(
+          size,
+          new ThreadFactoryBuilder()
+              .setDaemon(true)
+              .setPriority(Thread.NORM_PRIORITY)
+              .setNameFormat("iceberg-commit-table-pool-%d")
+              .build());
+    } else {
+      return null;
```

Comment:
just a thought: does it make sense to return a `Executors.singleThreadedExecutor` here? not sure if it make sense vs. passing in a null to the Tasks API, but it would simplify with the null checks when shutting the pool down

Reply:
Tasks API specifically handles `null` to run the stuff in the same thread as the code. I think it is even more efficient than the `Executors.singleThreadedExecutor`. What do you think?

Reply:
I was suspecting the null might be slightly more performant, but though the singleThreadedExecutor might help with code readability and avoiding NPEs. But I'm fine either way.

Reply:
But assuming most of our inserts will deal with a single table, using the more efficient solution makes more sense.